More datatypes for groupUniqArray() function [#METR-22071]

ludv1x opened the pull request:
Also added additional Arena\* argument for IAggregateFunction::add() function (it allows to make more sophisticated implementations).

ludv1x committed:
Added three experimental implementations for groupUniqArray() aggregate function. [#METR-22071]

ludv1x committed:
Leave only HashSet implementation. [#METR-22071]

ludv1x committed:
Added new Arena parameter to add() method of IAggregateFunction. [#METR-22071]

ludv1x committed:
Fixed dynamic complation errors. [#METR-22071]

alexey-milovidov commented on dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h:
```diff
@@ -70,7 +71,7 @@ class AggregateFunctionGroupUniqArray
 		size_t size = set.size();
 		writeVarUInt(size, buf);
-		for (typename State::Set::const_iterator it = set.begin(); it != set.end(); ++it)
+		for (auto it = set.begin(); it != set.end(); ++it)
```
Let go further and write `for (const auto & elem : set)`

alexey-milovidov commented on dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h:
```diff
@@ -94,11 +95,142 @@ class AggregateFunctionGroupUniqArray
+/// Generic implementation
+struct AggreagteFunctionGroupUniqArrayGenericData
+{
+	using Set = HashSetWithSavedHash<StringRef, StringRefHash, HashTableGrower<4>, HashTableAllocatorWithStackMemory<16>>;
```
Maybe there is inconsistency, because in HashTableGrower, parameter is power of two for space of hash table in number of elements, but in HashTableAllocatorWithStackMemory, parameter is number of bytes.
Please compare with other uses.

alexey-milovidov commented on dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h:
```diff
@@ -94,11 +95,142 @@ class AggregateFunctionGroupUniqArray
+	Set value;
+};
+
+template <bool is_plain_column=false>
```
Need whitespaces.

alexey-milovidov commented on dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h:
```diff
@@ -94,11 +95,142 @@ class AggregateFunctionGroupUniqArray
+template <bool is_plain_column=false>
+class AggreagteFunctionGroupUniqArrayGeneric : public IUnaryAggregateFunction<AggreagteFunctionGroupUniqArrayGenericData, AggreagteFunctionGroupUniqArrayGeneric<is_plain_column>>
+{
+	mutable DataTypePtr input_data_type;
```
Why mutable?

alexey-milovidov commented on dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h:
```diff
@@ -94,11 +95,142 @@ class AggregateFunctionGroupUniqArray
+		auto & set = this->data(place).value;
+		writeVarUInt(set.size(), buf);
+
+		for (auto & elem: set)
```
In all other places, we put whitespace in both sides of colon.

alexey-milovidov commented on dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h:
```diff
@@ -94,11 +95,142 @@ class AggregateFunctionGroupUniqArray
+		for (size_t i = 0; i < size; i++)
+		{
+			readStringBinary(str_buf, buf);
+			set.insert(StringRef(str_buf));
```
Dangling StringRef. (Data for strings is not owned.)
Need to test for this method with distributed query processing.

It's worth to add `readStringBinaryInto` method, that reads directly into Arena, by analogy with `readStringInto` method.

alexey-milovidov commented on dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h:
```diff
@@ -94,11 +95,142 @@ class AggregateFunctionGroupUniqArray
+
+		if (!is_plain_column)
+		{
+			if (!likely(inserted))
```
Non obvious, that is is likely.
I think, many repeated elements is frequent case too.
Better to remove `likely`.

In fact, benefit from `likely` is very rare and small.

alexey-milovidov commented on dbms/include/DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h:
```diff
@@ -94,11 +95,142 @@ class AggregateFunctionGroupUniqArray
+	{
+		auto & cur_set = this->data(place).value;
+		auto & rhs_set = this->data(rhs).value;
+		cur_set.merge(rhs_set);
```
Need to check very carefully, that Arena for rhs remains alive.
I don't remember, whether it is true.

alexey-milovidov commented on dbms/include/DB/AggregateFunctions/IAggregateFunction.h:
```diff
@@ -77,7 +78,7 @@ class IAggregateFunction
 
 	/// Добавить значение. columns - столбцы, содержащие аргументы, row_num - номер строки в столбцах.
-	virtual void add(AggregateDataPtr place, const IColumn ** columns, size_t row_num) const = 0;
+	virtual void add(AggregateDataPtr place, const IColumn ** columns, size_t row_num, Arena * arena) const = 0;
```
Comment was not edited to document `arena` parameter.

alexey-milovidov commented on dbms/include/DB/AggregateFunctions/IAggregateFunction.h:
```diff
@@ -144,6 +145,11 @@ class IAggregateFunctionHelper : public IAggregateFunction
 	{
 		return __alignof__(Data);
 	}
+
```
Was not removed.

alexey-milovidov commented on dbms/include/DB/Common/HashTable/HashTable.h:
```diff
@@ -70,6 +70,10 @@ bool check(const T x) { return x == 0; }
 template <typename T>
 void set(T & x) { x = 0; }
 
+bool check(const std::string & x);
```
Now useless?

alexey-milovidov commented on dbms/include/DB/Interpreters/Aggregator.h:
```diff
@@ -44,7 +44,7 @@ namespace ErrorCodes
   * Двухуровневая хэш-таблица работает чуть медленнее при маленьком количестве различных ключей,
   *  но при большом количестве различных ключей лучше масштабируется, так как позволяет
-  *  распараллелить некоторые операции (слияние, пост-обработку) естественным образом.
+  * распараллелить некоторые операции (слияние, пост-обработку) естественным образом.
```
Accidential modification.

alexey-milovidov commented on dbms/include/DB/Interpreters/Aggregator.h:
```diff
@@ -384,7 +384,7 @@ struct AggregationMethodConcat
 			/** Исправление, если все ключи - пустые массивы. Для них в хэш-таблицу записывается StringRef нулевой длины, но с ненулевым указателем.
 			  * Но при вставке в хэш-таблицу, такой StringRef оказывается равен другому ключу нулевой длины,
-			  *  у которого указатель на данные может быть любым мусором и использовать его нельзя.
+			  * у которого указатель на данные может быть любым мусором и использовать его нельзя.
```
Same as above.

alexey-milovidov commented on dbms/include/DB/Interpreters/Aggregator.h:
```diff
@@ -977,7 +977,7 @@ class Aggregator
 
 	/** Установить блок - пример результата,
-	  *  только если он ещё не был установлен.
+	  * только если он ещё не был установлен.
```
.

alexey-milovidov commented on dbms/src/AggregateFunctions/AggregateFunctionGroupUniqArray.cpp:
```diff
@@ -1,13 +1,36 @@
 #include <DB/AggregateFunctions/AggregateFunctionFactory.h>
 #include <DB/AggregateFunctions/AggregateFunctionGroupUniqArray.h>
 #include <DB/AggregateFunctions/Helpers.h>
+#include <common/logger_useful.h>
```
May remove.

alexey-milovidov commented on dbms/src/AggregateFunctions/AggregateFunctionGroupUniqArray.cpp:
```diff
@@ -1,13 +1,36 @@
+{
+		 if (typeid_cast<const DataTypeDateTime *>(&argument_type))	return new AggregateFunctionGroupUniqArray<UInt32>;
+	else if (typeid_cast<const DataTypeDate *>(&argument_type)) 	return new AggregateFunctionGroupUniqArray<UInt16>;
+	else
```
More clear is to use `DataTypeDateTime::FieldType` and `DataTypeDate::FieldType`.

alexey-milovidov commented on dbms/src/AggregateFunctions/AggregateFunctionGroupUniqArray.cpp:
```diff
@@ -1,13 +1,36 @@
+	{
+		/// Check that we can use plain version of AggreagteFunctionGroupUniqArrayGeneric
+		if (typeid_cast<const DataTypeString*>(&argument_type))
+			return new AggreagteFunctionGroupUniqArrayGeneric<true>;
```
And FixedString?

alexey-milovidov commented on dbms/src/Client/Benchmark.cpp:
```diff
@@ -58,11 +59,11 @@ class Benchmark
 			const String & host_, UInt16 port_, const String & default_database_,
 			const String & user_, const String & password_, const String & stage,
-			bool randomize_,
+			bool randomize_, size_t num_repetions_, double max_time_,
```
Typo.

alexey-milovidov commented on dbms/src/Client/Benchmark.cpp:
```diff
@@ -206,6 +209,18 @@ class Benchmark
+
+			if (num_repetions > 0 && info_total.queries >= num_repetions)
+			{
+				std::cout << "The execution is broken since request number of loops is reached\n";
```
In fact, iterations, not loops.
You may pass many queries to benchmark and it will loop many times through them.

alexey-milovidov commented on dbms/src/Client/Benchmark.cpp:
```diff
@@ -183,7 +186,7 @@ class Benchmark
 
 		/// В цикле, кладём все запросы в очередь.
-		for (size_t i = 0; !interrupt_listener.check(); ++i)
+		for (size_t i = 0; !interrupt_listener.check() && (!(num_repetions > 0) || i < num_repetions); ++i)
```
According to next three lines, you cannot specify `num_repetitions` greater than `queries.size()`.

alexey-milovidov commented on dbms/src/Client/Benchmark.cpp:
```diff
@@ -206,6 +209,18 @@ class Benchmark
+
+			if (max_time > 0 && info_total.watch.elapsedSeconds() >= max_time)
+			{
+				std::cout << "The execution is broken since requested time limit is reached\n";
```
broken means "сломано".

alexey-milovidov commented on dbms/src/Client/Benchmark.cpp:
```diff
@@ -346,13 +361,15 @@ int main(int argc, char ** argv)
 			("help", "produce help message")
 			("concurrency,c", boost::program_options::value<unsigned>()->default_value(1), "number of parallel queries")
-			("delay,d", boost::program_options::value<double>()->default_value(1), "delay between reports in seconds")
+			("delay,d", boost::program_options::value<double>()->default_value(1), "delay between reports in seconds (set 0 to disable)")
```
Unclear. To disable reports or to disable delay?

alexey-milovidov commented on dbms/src/Client/Benchmark.cpp:
```diff
@@ -346,13 +361,15 @@ int main(int argc, char ** argv)
 			("password", boost::program_options::value<std::string>()->default_value(""), "")
 			("database", boost::program_options::value<std::string>()->default_value("default"), "")
 			("stage", boost::program_options::value<std::string>()->default_value("complete"), "request query processing up to specified stage")
+			("loops,l", boost::program_options::value<size_t>()->default_value(0), "number of tests repetions")
```
Don't forget about typo.

alexey-milovidov commented on dbms/src/Client/Benchmark.cpp:
```diff
@@ -346,13 +361,15 @@ int main(int argc, char ** argv)
 			("database", boost::program_options::value<std::string>()->default_value("default"), "")
 			("stage", boost::program_options::value<std::string>()->default_value("complete"), "request query processing up to specified stage")
+			("loops,l", boost::program_options::value<size_t>()->default_value(0), "number of tests repetions")
+			("timelimit,t", boost::program_options::value<double>()->default_value(0.), "stop repeating after specified time limit")
```
If many queries, it means not 'repeating' but 'running'.

ludv1x replied:
I deleted Benchmark.cpp changes from this PR.

There are some additional problems with current version of the benchmark.
I will resolve them and make separated PR.

alexey-milovidov replied:
Ok.

alexey-milovidov commented on dbms/src/Interpreters/Aggregator.cpp:
```diff
@@ -431,7 +432,7 @@ AggregatedDataVariants::Type Aggregator::chooseAggregationMethod(const ConstColu
 
 
-void Aggregator::createAggregateStates(AggregateDataPtr & aggregate_data) const
+void Aggregator::createAggregateStates(AggregateDataPtr & aggregate_data, Arena * arena) const
```
No need for `arena` in this method.

alexey-milovidov commented on dbms/src/Interpreters/Aggregator.cpp:
```diff
@@ -1003,6 +1005,7 @@ void NO_INLINE Aggregator::convertToBlockImplFinal(
 	ColumnPlainPtrs & final_aggregate_columns,
 	const Sizes & key_sizes) const
 {
+	//LOG_DEBUG(log, "convertToBlockImplFinal start");
```
Was not removed.

alexey-milovidov commented on dbms/src/Interpreters/Aggregator.cpp:
```diff
@@ -1808,6 +1830,8 @@ void NO_INLINE Aggregator::mergeStreamsImplCase(
 				value + offsets_of_aggregate_states[j],
 				(*aggregate_columns[j])[i]);
+
+		LOG_DEBUG(log, "mergeStreamsImplCase");
```
.

ludv1x committed:
Added requested changes. [#METR-22071]

ludv1x committed:
Added arena parameter into merge() method of aggregation function states. [#METR-22071]

ludv1x committed:
Added simple test for groupUniqArray(String). [#METR-22071]

Also reverted Benchmark.cpp it will be added in separate PR.

alexey-milovidov commented on dbms/include/DB/Functions/FunctionsMiscellaneous.h:
```diff
@@ -1245,7 +1245,7 @@ class FunctionRunningAccumulate : public IFunction
 		for (const auto & state_to_add : states)
 		{
-			agg_func.merge(place.get(), state_to_add);
+			agg_func.merge(place.get(), state_to_add, nullptr); /// Empty arena!
```
Need to create Arenas. You may use ObjectPool (our ObjectPool, not from Poco) for this task. See usages.

alexey-milovidov replied:
Or add 'bool allocatesMemoryInArena()' method to IAggregateFunction and throw exception for that functions here.

alexey-milovidov commented on dbms/src/Interpreters/Aggregator.cpp:
```diff
@@ -1476,9 +1488,13 @@ void NO_INLINE Aggregator::mergeBucketImpl(
 		AggregatedDataVariants & current = *data[i];
 
+		/// Select Arena to avoid race conditions
+		Arena * arena = res->aggregates_pools.at(static_cast<size_t>(bucket) % size).get();
```
Need to create more aggregates_pools if has less than max_threads of them.

ludv1x committed:
Added allocatesMemoryInArena() method for aggregate functions.

Fixed runningAccumulate, now it works properly for complex functions.
More accurate threads handling in Aggregator.

alexey-milovidov commented on dbms/tests/queries/0_stateless/00373_group_uniq_array_simple.sql:
```diff
@@ -0,0 +1,10 @@
+INSERT INTO test.group_uniq SELECT 5 as id, toString(number % 100) as v FROM system.numbers LIMIT 10000000;
+
+SELECT length(groupUniqArray(v)) FROM test.group_uniq GROUP BY id ORDER BY id;
+--SELECT length(groupUniqArray(v)) FROM remote('127.0.0.{1,2,3,4}', 'test', 'group_uniq') GROUP BY id ORDER BY id;
```
Why test for distributed is commented?

ludv1x replied:
It isn't obvious that two sequential SELECT calls produce output only for first SELECT.
I was lazy to add an two similar test.
Now I use UNION ALL.

alexey-milovidov replied:
It is not true. Look at most of other .sql tests.
All sql tests are run in `multiquery` mode.

alexey-milovidov replied:
But when you running query with UNION ALL, you get non-deterministic order of result, when max_threads != 1. So, test will be flappy. Please refine.
You may send another pull request.

alexey-milovidov commented:
Need to add test for array of numbers and array of strings types of arguments.

ludv1x committed:
Added more tests for groupUniqArray() function. [#METR-22071]

ludv1x commented:
Ok
